### PR TITLE
Adding sparse possibilities

### PR DESCRIPTION
**Pull request by teytaud**
## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Motivation and Context / Related issue

Our colleagues from GP got really cool results in explainable AI. 
Let's see if we can do something similar with sparse neural nets.

## Checklist

- [x] The documentation is up-to-date with the changes I made.
- [x] I have read the [**CONTRIBUTING**](https://facebookresearch.github.io/nevergrad/contributing.html) document and completed the CLA (see [**CLA**](https://facebookresearch.github.io/nevergrad/contributing.html#contributor-license-agreement-cla)).
- [x] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

File: nevergrad/optimization/optimizerlib.py [hunk 1]
```diff
@@ -247,6 +255,12 @@ def _internal_ask_candidate(self) -> p.Parameter:
                     "portfolio": mutator.portfolio_discrete_mutation,
                 }[mutation]
                 data = func(pessimistic_data, arity=self.arity_for_discrete_mutation)
+            if self.sparse:
+                data = np.asarray(data)
+                zeroing = (
+                    np.asarray(np.random.randint(data.size + 1, size=data.size).reshape(data.shape)) == 0
+                )
```
Comment (jrapin) [inline nevergrad/optimization/optimizerlib.py:262]: Why use asarray? It should already be one
```suggestion
                zeroing = np.random.randint(data.size + 1, size=data.size).reshape(data.shape)) == 0
```
Also, is it intentional that in some cases nothing will change?

File: nevergrad/optimization/optimizerlib.py [hunk 2]
```diff
@@ -330,13 +344,17 @@ def __init__(
         crossover: bool = False,
         rotation: bool = False,
         use_pareto: bool = False,
+        sparse: bool = False,
```
Comment (jrapin) [inline nevergrad/optimization/optimizerlib.py:347]: please update the docstring to explain what it does and when to use it

File: nevergrad/optimization/optimizerlib.py [hunk 3]
```diff
@@ -330,13 +344,17 @@ def __init__(
...
     ) -> None:
         super().__init__(_OnePlusOne, locals())
 
 
 OnePlusOne = ParametrizedOnePlusOne().set_name("OnePlusOne", register=True)
 NoisyOnePlusOne = ParametrizedOnePlusOne(noise_handling="random").set_name("NoisyOnePlusOne", register=True)
 DiscreteOnePlusOne = ParametrizedOnePlusOne(mutation="discrete").set_name("DiscreteOnePlusOne", register=True)
+SparseDiscreteOnePlusOne = ParametrizedOnePlusOne(mutation="discrete", sparse=True).set_name(
+    "SparseDiscreteOnePlusOne", register=True
+)
```
Comment (jrapin) [inline nevergrad/optimization/optimizerlib.py:357]: can we keep them to experimental variants for now?

File: nevergrad/optimization/optimizerlib.py [hunk 4]
```diff
@@ -247,6 +255,10 @@ def _internal_ask_candidate(self) -> p.Parameter:
                     "portfolio": mutator.portfolio_discrete_mutation,
                 }[mutation]
                 data = func(pessimistic_data, arity=self.arity_for_discrete_mutation)
+            if self.sparse:
+                data = np.asarray(data)
+                zeroing = np.random.randint(data.size + 1, size=data.size).reshape(data.shape) == 0
```
Comment (jrapin) [inline nevergrad/optimization/optimizerlib.py:260]:
```suggestion
                zeroing = self._rng.randint(data.size + 1, size=data.size).reshape(data.shape) == 0
```
fixes reproducibility

Comment (teytaud) [reply nevergrad/optimization/optimizerlib.py]: Thx!

**Pull request merged**